File: three_values_solution.py
"""Exact solver for the original problem with values 1, 2, and 3.

Input:  n
Output: maximum possible sum of an n by n grid

The algorithm is a frontier dynamic program.  Each frontier cell stores its
value and which required neighbor values it has already seen.
"""

from __future__ import annotations

import sys


# Frontier state for one cell:
#   0: value 1
#   1: value 2, has not seen a neighboring 1
#   2: value 2, has seen a neighboring 1
#   3: value 3, has seen neither 1 nor 2
#   4: value 3, has seen 1 only
#   5: value 3, has seen 2 only
#   6: value 3, has seen both 1 and 2
#   7: no cell (used only before the first row)

CELL_VALUE = (1, 2, 2, 3, 3, 3, 3, 0)
IS_VALID = (True, False, True, False, False, False, True, False)


def add_neighbor(state: int, neighbor_value: int) -> int:
    """Update a frontier state after exposing one adjacent cell."""
    if state == 0 or state == 7:
        return state

    if state <= 2:
        return 2 if neighbor_value == 1 else state

    flags = state - 3
    if neighbor_value == 1:
        flags |= 1
    elif neighbor_value == 2:
        flags |= 2
    return 3 + flags


# Small transition tables save considerable work in the inner DP loop.
ADD_NEIGHBOR = tuple(
    tuple(add_neighbor(state, value) for value in range(4))
    for state in range(8)
)


def make_state(value: int, upper_value: int, left_value: int) -> int:
    state = 0 if value == 1 else 1 if value == 2 else 3
    if upper_value:
        state = ADD_NEIGHBOR[state][upper_value]
    if left_value:
        state = ADD_NEIGHBOR[state][left_value]
    return state


MAKE_STATE = tuple(
    tuple(
        tuple(make_state(value, upper, left) for left in range(4))
        for upper in range(4)
    )
    for value in range(4)
)


def reverse_frontier(frontier: int, width: int) -> int:
    """Reflect a complete row to merge horizontally symmetric states."""
    result = 0
    for column in range(width):
        digit = (frontier >> (3 * column)) & 7
        result |= digit << (3 * (width - 1 - column))
    return result


def maximum_sum(n: int) -> int:
    if n <= 0:
        raise ValueError("n must be positive")

    empty_frontier = sum(7 << (3 * column) for column in range(n))
    dp: dict[int, int] = {empty_frontier: 0}

    for _row in range(n):
        for column in range(n):
            next_dp: dict[int, int] = {}
            shift = 3 * column
            current_mask = 7 << shift
            left_shift = shift - 3
            left_mask = 0 if column == 0 else 7 << left_shift

            for frontier, old_sum in dp.items():
                upper_state = (frontier >> shift) & 7
                left_state = 7 if column == 0 else (frontier >> left_shift) & 7
                upper_value = CELL_VALUE[upper_state]
                left_value = CELL_VALUE[left_state]

                for value in (1, 2, 3):
                    # The new cell is the final possible neighbor of the cell
                    # above.  That upper cell must now satisfy its rule.
                    if upper_state != 7:
                        finished_upper = ADD_NEIGHBOR[upper_state][value]
                        if not IS_VALID[finished_upper]:
                            continue

                    new_left = ADD_NEIGHBOR[left_state][value]
                    new_current = MAKE_STATE[value][upper_value][left_value]

                    new_frontier = (
                        (frontier & ~current_mask) | (new_current << shift)
                    )
                    if column != 0:
                        new_frontier = (
                            (new_frontier & ~left_mask)
                            | (new_left << left_shift)
                        )

                    # Once a row is complete, its reflection has an identical
                    # remaining subproblem, so retain one orientation only.
                    if column == n - 1:
                        new_frontier = min(
                            new_frontier,
                            reverse_frontier(new_frontier, n),
                        )

                    new_sum = old_sum + value
                    previous = next_dp.get(new_frontier)
                    if previous is None or new_sum > previous:
                        next_dp[new_frontier] = new_sum

            dp = next_dp

    answer = 0
    for frontier, grid_sum in dp.items():
        if all(
            IS_VALID[(frontier >> (3 * column)) & 7]
            for column in range(n)
        ):
            answer = max(answer, grid_sum)
    return answer


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if len(data) != 1:
        raise SystemExit("expected one positive integer n")

    n = int(data[0])
    print(maximum_sum(n))


if __name__ == "__main__":
    main()

File: three_values_solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

// Frontier state for one cell:
//   0       : value 1
//   1, 2    : value 2, without/with an adjacent 1
//   3..6    : value 3, plus two bits saying whether adjacent 1 and 2 exist
//   7       : empty (only used before the first row is filled)

int cellValue(int state) {
    if (state == 0) return 1;
    if (state <= 2) return 2;
    return 3;
}

int addNeighbor(int state, int neighborValue) {
    if (state == 0) return state;  // A 1-cell has no condition.

    if (state <= 2) {
        return neighborValue == 1 ? 2 : state;
    }

    int flags = state - 3;
    if (neighborValue == 1) flags |= 1;
    if (neighborValue == 2) flags |= 2;
    return 3 + flags;
}

bool isValid(int state) {
    return state == 0                 // value 1
        || state == 2                 // value 2 with an adjacent 1
        || state == 6;                // value 3 with adjacent 1 and 2
}

int makeState(int value, int upperValue, int leftValue) {
    int state = (value == 1 ? 0 : value == 2 ? 1 : 3);
    if (upperValue != 0) state = addNeighbor(state, upperValue);
    if (leftValue != 0) state = addNeighbor(state, leftValue);
    return state;
}

struct Parent {
    uint64_t previousFrontier;
    int value;
};

struct Result {
    int maximumSum;
    vector<string> grid;
};

Result solve(int n) {
    // Each frontier cell occupies three bits because it has eight states.
    uint64_t emptyFrontier = 0;
    for (int col = 0; col < n; ++col) {
        emptyFrontier |= 7ULL << (3 * col);
    }

    // frontier encoding -> maximum sum obtained so far
    unordered_map<uint64_t, int> dp, nextDp;
    dp[emptyFrontier] = 0;

    // parents[step][frontier] tells us which state and value produced this
    // optimal frontier at that step.
    vector<unordered_map<uint64_t, Parent>> parents;
    parents.reserve(n * n);

    for (int row = 0; row < n; ++row) {
        for (int col = 0; col < n; ++col) {
            nextDp.clear();
            nextDp.reserve(dp.size() * 2);
            unordered_map<uint64_t, Parent> stepParents;
            stepParents.reserve(dp.size() * 2);

            const int shift = 3 * col;
            const uint64_t currentMask = 7ULL << shift;
            const uint64_t leftMask =
                (col == 0 ? 0 : 7ULL << (shift - 3));

            for (const auto& [frontier, sum] : dp) {
                int upperState = static_cast<int>((frontier >> shift) & 7);
                int leftState = (col == 0)
                    ? 7
                    : static_cast<int>((frontier >> (shift - 3)) & 7);

                int upperValue =
                    (upperState == 7 ? 0 : cellValue(upperState));
                int leftValue =
                    (leftState == 7 ? 0 : cellValue(leftState));

                for (int value = 1; value <= 3; ++value) {
                    // The new cell is the last unknown neighbor of the upper
                    // cell, so the upper cell must now satisfy its condition.
                    if (upperState != 7
                        && !isValid(addNeighbor(upperState, value))) {
                        continue;
                    }

                    int newLeftState = (leftState == 7)
                        ? 7
                        : addNeighbor(leftState, value);
                    int newCurrentState =
                        makeState(value, upperValue, leftValue);

                    uint64_t newFrontier =
                        (frontier & ~currentMask)
                        | (static_cast<uint64_t>(newCurrentState) << shift);

                    if (col != 0) {
                        newFrontier =
                            (newFrontier & ~leftMask)
                            | (static_cast<uint64_t>(newLeftState)
                               << (shift - 3));
                    }

                    int newSum = sum + value;
                    auto it = nextDp.find(newFrontier);
                    if (it == nextDp.end() || newSum > it->second) {
                        nextDp[newFrontier] = newSum;
                        stepParents[newFrontier] = {frontier, value};
                    }
                }
            }

            dp.swap(nextDp);
            parents.push_back(move(stepParents));
        }
    }

    // Bottom-row cells have no lower neighbor, so validate them here.
    int answer = 0;
    uint64_t bestFrontier = 0;
    for (const auto& [frontier, sum] : dp) {
        bool validGrid = true;
        for (int col = 0; col < n; ++col) {
            int state = static_cast<int>((frontier >> (3 * col)) & 7);
            if (!isValid(state)) {
                validGrid = false;
                break;
            }
        }
        if (validGrid && sum > answer) {
            answer = sum;
            bestFrontier = frontier;
        }
    }

    // Follow the stored parents backwards to recover one optimal grid.
    vector<string> grid(n, string(n, '0'));
    uint64_t frontier = bestFrontier;
    for (int step = n * n - 1; step >= 0; --step) {
        const Parent& parent = parents[step].at(frontier);
        grid[step / n][step % n] = static_cast<char>('0' + parent.value);
        frontier = parent.previousFrontier;
    }

    return {answer, grid};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n) || n < 1 || n > 20) {
        cerr << "Expected 1 <= n <= 20.\n";
        return 1;
    }

    Result result = solve(n);
    cout << "maximum sum = " << result.maximumSum << '\n';
    cout << "arrangement:\n";
    for (const string& row : result.grid) {
        cout << row << '\n';
    }
    return 0;
}
